### PR TITLE
stream: process more than one write req per loop tick

### PR DESCRIPTION
**Pull request by ywave620**
Resolve https://github.com/libuv/libuv/issues/3668

### CODE REVIEW

File: src/unix/stream.c [hunk 1]
```diff
@@ -905,10 +912,13 @@ static void uv__write(uv_stream_t* stream) {
...
         uv__write_req_finish(req);
-        return;  /* TODO(bnoordhuis) Start trying to write the next request. */
+        if (cnt-- > 0) {
+          continue; /* Start trying to write the next request. */
+        }
+        return;
       }
     } else if (n != UV_EAGAIN)
-      break;
+      goto error;
```
Comment (santigimeno) [inline src/unix/stream.c:921]: Why this change?

File: test/test-tcp-write-in-a-row.c [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+  }
+
+  free(w);
+}
+
+static void connect_cb(uv_connect_t *_, int status)
+{
+  int r;
+  uv_buf_t buf;
+  uv_write_t *req;
```
Comment (santigimeno) [inline test/test-tcp-write-in-a-row.c:66]: Instead of allocating 2 `uv_write_t` requests in the heap and handling their lifetime, you could have 2 global `uv_write_t` so the code is simplified

**Review by santigimeno — APPROVED** [review]
LGTM with some nits.

**Comment by santigimeno** [comment]
BTW I think this should be created against the `v1.x` branch

File: src/unix/stream.c (hunk 1 above)
Comment (ywave620) [reply src/unix/stream.c:921]: I think labeling `error` to the last few lines of this function just makes thing more clear. The original function mixs `return` and `break` in a large loop body, one has to tell the semantic difference between these two control statements.

File: src/unix/stream.c [hunk 3]
```diff
@@ -884,9 +884,16 @@ static void uv__write(uv_stream_t* stream) {
...
   ssize_t n;
+  int cnt;
 
   assert(uv__stream_fd(stream) >= 0);
 
+  /* Prevent loop starvation when the consumer of this stream read as fast as (or faster than)
+   * we can write it. This `cnt` mechanism does not need to change even we switch to edge-triggered
+   * I/O.
+   */
+  cnt = 32;
```
Comment (bnoordhuis) [inline src/unix/stream.c:895]: Can you name this `count` for consistency with `uv__read()` from this file?

File: src/unix/stream.c [hunk 4]
```diff
@@ -905,10 +912,13 @@ static void uv__write(uv_stream_t* stream) {
       req->send_handle = NULL;
       if (uv__write_req_update(stream, req, n)) {
         uv__write_req_finish(req);
-        return;  /* TODO(bnoordhuis) Start trying to write the next request. */
+        if (cnt-- > 0) {
+          continue; /* Start trying to write the next request. */
+        }
```
Comment (bnoordhuis) [inline src/unix/stream.c:917]:
```suggestion
        if (cnt-- > 0)
          continue;  /* Start trying to write the next request. */
```

File: test/test-tcp-write-in-a-row.c [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+
+static uv_tcp_t server;
+static uv_tcp_t client;
+static uv_tcp_t incoming;
+static int connect_cb_called;
+static int close_cb_called;
+static int connection_cb_called;
+static int write_cb_called;
+
+// 10 MB, which is large than the send buffer size and the recv buffer
```
Comment (bnoordhuis) [inline test/test-tcp-write-in-a-row.c:37]: C89 comments, please, here and elsewhere.
```suggestion
/* 10 MB, which is large than the send buffer size and the recv buffer */
```

File: test/test-tcp-write-in-a-row.c [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+{
+  int r;
+  uv_buf_t buf;
+  uv_write_t *req;
+  size_t write_queue_size0, write_queue_size1;
+
+  ASSERT(status == 0);
+  connect_cb_called++;
+
+  req = (uv_write_t *)malloc(sizeof(uv_write_t));
```
Comment (bnoordhuis) [inline test/test-tcp-write-in-a-row.c:72]:
```suggestion
  req = malloc(sizeof(*req));
  ASSERT_NOT_NULL(req);
```
But as Santiago said, simpler to use static storage.

File: test/test-tcp-write-in-a-row.c [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+
+  // fire a big write
+  buf = uv_buf_init(data, sizeof(data));
+  r = uv_write(req, (uv_stream_t *)&client, &buf, 1, write_cb);
```
Comment (bnoordhuis) [inline test/test-tcp-write-in-a-row.c:76]: Style: star leans left (here and elsewhere)
```suggestion
  r = uv_write(req, (uv_stream_t*)&client, &buf, 1, write_cb);
```

**Review by bnoordhuis — COMMENTED** [review]
Thanks for the PR. Some style nits but the general thrust looks good.

File: src/unix/stream.c (hunk 3 above)
Comment (ywave620) [reply src/unix/stream.c:895]: > BTW I think this should be created against the `v1.x` branch

The CONTRIBUTING.md says `As a rule of thumb, bug fixes go into the stable branch
while new features go into the master branch. The stable branch is effectively frozen; patches that change the libuv
API/ABI or affect the run-time behavior of applications get rejected`

**Comment by ywave620** [comment]
> BTW I think this should be created against the `v1.x` branch

The CONTRIBUTING.md says `As a rule of thumb, bug fixes go into the stable branch
while new features go into the master branch. The stable branch is effectively frozen; patches that change the libuv
API/ABI or affect the run-time behavior of applications get rejected`

**Comment by bnoordhuis** [comment]
We should probably reword that to make our policy clearer. Backwards incompatible changes indeed cannot land in the v1.x branch but backwards compatible changes (which I would classify this as) and new APIs can.

**Comment by ywave620** [comment]
Make change on v1.x rather than master. https://github.com/libuv/libuv/pull/3728
